src/fastpath/fastpath.c

```diff
@@ -6,15 +6,175 @@
+    if (unlikely(!cap_notification_cap_get_capNtfnCanSend(cap))) {
+        slowpath(SysSend);
+    }
+
+    /* Get the notification address */
+    notification_t *ntfnPtr = NTFN_PTR(cap_notification_cap_get_capNtfnPtr(cap));
+
+    /* Get the notification state */
+    uint32_t ntfnState = notification_ptr_get_state(ntfnPtr);
+    tcb_t *dest;
+
+    if (ntfnState == NtfnState_Waiting) {
+        /* get the destination thread */
+        dest = TCB_PTR(notification_ptr_get_ntfnQueue_head(ntfnPtr));
+    } else {
+        /* get the bound tcb */
+        dest = (tcb_t *) notification_ptr_get_ntfnBoundTCB(ntfnPtr);
+    }
+
+    /* Get the notification badge */
+    word_t badge = cap_notification_cap_get_capNtfnBadge(cap);
+    switch (ntfnState) {
+    case NtfnState_Active: {
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+        ksKernelEntry.is_fastpath = true;
+#endif
+        ntfn_set_active(ntfnPtr, badge | notification_ptr_get_ntfnMsgIdentifier(ntfnPtr));
+        restore_user_context();
+        UNREACHABLE();
+    }
+    case NtfnState_Idle: {
+        /* Check if we are bound and that thread is waiting for a message */
+        if (dest && thread_state_ptr_get_tsType(&dest->tcbState) == ThreadState_BlockedOnReceive) {
+            /* Signal to higher prio thread is NOT fastpathed*/
+            if (NODE_STATE(ksCurThread)->tcbPriority < dest->tcbPriority) {
+                slowpath(SysSend);
+            }
+
+            /* Basically equivalent to maybeDonateSchedContext. Check whether the thread already has
+             * a SC  or if one can be donated from the notification. If neither is true, go to
+             * slowpath */
+            if (!dest->tcbSchedContext) {
+                sc = SC_PTR(notification_ptr_get_ntfnSchedContext(ntfnPtr));
+                if (sc == NULL || sc->scTcb != NULL) {
+                    slowpath(SysSend);
+                    UNREACHABLE();
+                }
+            }
+
+            /*  Point of no return */
+
+            /* Equivalent to cancel_ipc */
+            endpoint_t *ep_ptr;
+            ep_ptr = EP_PTR(thread_state_get_blockingObject(dest->tcbState));
+            endpoint_ptr_set_epQueue_head_np(ep_ptr, TCB_REF(dest->tcbEPNext));
+            if (unlikely(dest->tcbEPNext)) {
+                dest->tcbEPNext->tcbEPPrev = NULL;
+            } else {
+                endpoint_ptr_mset_epQueue_tail_state(ep_ptr, 0, EPState_Idle);
+            }
+
+
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+            ksKernelEntry.is_fastpath = true;
+#endif
+            if (!dest->tcbSchedContext) {
+                schedContext_donate(sc, dest);
+            }
+            assert(dest->tcbSchedContext);
+
+            /* Enqueue signalled but continue executing signaller */
+            setRegister(dest, badgeRegister, badge);
+            thread_state_ptr_set_tsType_np(&dest->tcbState, ThreadState_Running);
+            tcbSchedEnqueue(dest);
```

Comment:
Yeah, I think this makes more sense.

Reply:
Coming back to this, I think the correct behavior to have equivalency with the slowpath is actually to enqueue. The slowpath uses possibleSwitchTo() for the dest thread, which, in combination with `schedule()`, seems to enqueue the destination thread (in the case that it is lower priority than the current thread). I talked to Kent a bit about this to understand the motivation behind this and I can forward the email to you if you would like.

Reply:
Please do, this is unexpected behaviour and I would like to know why it's done.

Reply:
Done. The correct implementation might be a sort of mix of the two where lower priority threads are enqueued while equal priority threads are appended, which I think is the behaviour of the slowpath, at least for the non cross-node case. For cross-node, it seems like the destination thread is always enqueued by possibleSwitchTo().

Reply:
Thanks for the forward. To summarize, enqueue is used to optimise the case where a high priority server replies to a lower priority client. Basically higher priority tasks are allowed to influence the release queue order of lower priority tasks. This makes sense in general.

I'm not convinced this makes sense for notifications though. If I understood correctly, this means that if a high priority task notifies multiple tasks, they will be woken up in reverse order. For network packets this means that the last received packets will be handled first instead of the oldest.

Reply:
I think you're right about this behavior being weird, but after speaking to Gernot, the course of action seems to be that we should have equivalency with whatever the slowpath has for now but eventually revise the slowpath to a model that is more aligned with something like what you said (will probably involve a decent bit of shuffling stuff around) and adjust the fastpath as needed.

Reply:
Agreed, the fast path needs to follow the slow path, even if it doesn't make sense. I'll have a look at this issue next week or so.

Reply:
I'll mark this as resolved for this PR, since we have a separate issue for this open (#902).